src/common/utils/pagination.ts

```diff
@@ -0,0 +1,55 @@
+import { List, PaginationOptions } from '../interfaces';
+
+export class AutoPaginatable<T> {
+  readonly object: 'list' = 'list';
+  readonly options: PaginationOptions;
+
+  constructor(
+    private list: List<T>,
+    private apiCall: (params: PaginationOptions) => Promise<List<T>>,
+    options?: PaginationOptions,
+  ) {
+    this.options = {
+      ...options,
+    };
+  }
+
+  get data(): T[] {
+    return this.list.data;
+  }
+
+  get listMetadata() {
+    return this.list.listMetadata;
+  }
+
+  private async *generatePages(params: PaginationOptions): AsyncGenerator<T[]> {
```

Comment:
Commenting for posterity: we should look into exposing this via [`Symbol.asyncIterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/asyncIterator) if later developers decide they'd like to lazily iterate over the pages directly.

Until then, only exposing `autoPagination` seems reasonable.